src/software/ai/hl/stp/play/enemy_free_kick/enemy_free_kick_play_fsm.cpp

```diff
@@ -0,0 +1,189 @@
+#include "software/ai/hl/stp/tactic/crease_defender/crease_defender_tactic.h"
+#include "software/ai/hl/stp/tactic/move/move_tactic.h"
+#include "software/ai/hl/stp/tactic/pass_defender/pass_defender_tactic.h"
+#include "software/geom/algorithms/distance.h"
+#include "software/util/generic_factory/generic_factory.h"
+
+EnemyFreeKickPlayFSM::EnemyFreeKickPlayFSM(TbotsProto::AiConfig ai_config)
+    : ai_config(ai_config), crease_defenders({}), pass_defenders({})
+{
+}
+
+void EnemyFreeKickPlayFSM::setupEnemyKickerStrategy(const Update& event)
+{
+    setTactics(event, event.common.num_tactics);
+}
+
+void EnemyFreeKickPlayFSM::setTactics(const Update& event, unsigned int num_tactics)
+{
+    // One tactic is always designated as the free kick defender
+    unsigned int num_defenders             = num_tactics - 1;
+    PriorityTacticVector tactics_to_return = {{}, {}, {}};
+    Point block_kick_point;
+
+    auto enemy_threats = getAllEnemyThreats(
+        event.common.world_ptr->field(), event.common.world_ptr->friendlyTeam(),
+        event.common.world_ptr->enemyTeam(), event.common.world_ptr->ball(), false);
+
+    auto assignments = getAllDefenderAssignments(
+        enemy_threats, event.common.world_ptr->field(), event.common.world_ptr->ball(),
+        ai_config.defense_play_config().defender_assignment_config());
+
+    if (assignments.size() == 0)
+    {
+        return;
+    }
+
+    // Adds designated free kick defender to block the direction the kicker is facing
+    if (!enemy_threats.empty())
+    {
+        auto block_free_kicker = std::make_shared<PassDefenderTactic>();
+        Vector block_direction =
+            Vector::createFromAngle(enemy_threats[0].robot.orientation());
+        block_kick_point = event.common.world_ptr->ball().position() +
+                           block_direction.normalize(0.6 + 2 * ROBOT_MAX_RADIUS_METERS);
+        block_free_kicker->updateControlParams(block_kick_point);
+        tactics_to_return[0].push_back(block_free_kicker);
+    }
+
+    // Choose which defender assignments to assign defenders to based on number
+    // of tactics available to set after assigning kick off defender
+    std::vector<DefenderAssignment> crease_defender_assignments;
+    std::vector<DefenderAssignment> pass_defender_assignments;
+    unsigned int assigns_skipped = 0;
+    for (unsigned int i = 0; i < num_defenders; i++)
+    {
+        DefenderAssignment defender_assignment;
+        if (i < assignments.size())
+        {
+            defender_assignment = assignments.at(i + assigns_skipped);
+            defender_assignment = assignments.at(i);
+
+            while (defender_assignment.type == PASS_DEFENDER &&
+                   distance(defender_assignment.target, block_kick_point) <= 0.5)
+            {
+                if (i + assigns_skipped + 1 >= assignments.size())
+                {
+                    defender_assignment = assignments.front();
+                    break;
+                }
```

Comment:
we should push these assignments into a queue, and then pull from this queue on line 85.